sklearn/model_selection/_prediction.py

```diff
@@ -0,0 +1,339 @@
+                )
+            self._method = next(
+                (m for m, i in zip(supported_methods, has_methods) if i), None
+            )
+        else:
+            if self.method not in supported_methods:
+                raise ValueError(
+                    f"'method' should be one of {', '.join(supported_methods)}"
+                    f". Got {self.method} instead."
+                )
+            elif not hasattr(self.base_estimator, self.method):
+                raise TypeError(
+                    f"'base_estimator' does not implement {self.method}."
+                )
+            self._method = self.method
+        if (self.objective_metric not in ("tpr", "tnr", "cost_matrix") and
+                self.objective_value is not None):
+            raise ValueError(
+                f"When 'objective_metric' is a scoring function, "
+                f"'objective_value' should be None. Got "
+                f"{self.objective_metric} instead."
+            )
+        elif self.objective_metric == "cost_matrix":
+            if self.objective_value.shape != (2, 2):
+                raise ValueError(
+                    f"When 'objective_metric' is a cost matrix, it must be of "
+                    f"shape (2, 2). Got {self.objective_value.shape} instead."
+                )
+
+        # ensure binary classification if `pos_label` is not specified
+        # `classes.dtype.kind` in ('O', 'U', 'S') is required to avoid
+        # triggering a FutureWarning by calling np.array_equal(a, b)
+        # when elements in the two arrays are not comparable.
+        if (self.pos_label is None and (
+                self.classes_.dtype.kind in ('O', 'U', 'S') or
+                not (np.array_equal(self.classes_, [0, 1]) or
+                     np.array_equal(self.classes_, [-1, 1]) or
+                     np.array_equal(self.classes_, [0]) or
+                     np.array_equal(self.classes_, [-1]) or
+                     np.array_equal(self.classes_, [1])))):
+            classes_repr = ", ".join(repr(c) for c in self.classes_)
+            raise ValueError(
+                f"'y_true' takes value in {classes_repr} and 'pos_label' is "
+                f"not specified: either make 'y_true' take value in "
+                "{{0, 1}} or {{-1, 1}} or pass pos_label explicitly."
+            )
+        elif self.pos_label is None:
+            self._pos_label = 1
+        else:
+            self._pos_label = self.pos_label
+
+    @staticmethod
+    def _validate_data(X, y):
+        y = check_array(y, ensure_2d=False, dtype=None)
+        check_classification_targets(y)
+        y_type = type_of_target(y)
+        if y_type != 'binary':
+            raise ValueError(f'Expected target of binary type. Got {y_type}.')
+        return X, y
+
+    @staticmethod
+    def cost_sensitive_score(y_true, y_pred, cost_matrix):
+        cm = confusion_matrix(y_true, y_pred) * cost_matrix
+        return np.diag(cm) / cm.sum()
```

Comment:
Why pass a full cost matrix if you then only consider the diagonal?

We could define an arbitrary business-specific utility function: such as:

- gain for true positive
- gain (negative cost) for false positives
- gain (negative cost) for false negatives
- gain for true negative (generally 0 for imbalanced classification problems).

I which case `cost_senstive_score` would be `(confusion_matrix * gain_matrix).sum() / n_samples` or alternatively `-(confusion_matrix * cost_matrix).sum() / n_samples`.

WDYT?

This work could be later extended to have parametrized cost sensitive scorer objects that could both be used for optimal cutoff selection of the final classifier of a pipeline but also used by the SearchCV object for hyper-parameter tuning of the full pipeline in general.

Reply:
> Why pass a full cost matrix if you then only consider the diagonal?

`cm.sum()` will consider off-diagonal. If you penalize differently false-positive and false-negative, the cost will change in this case, isn't it?

Reply:
Indeed. This is a really hard to interpret how gains and costs are counteracts one another. I prefer my simple purely additive parametrization unless you can show me that yours make more sense.

Reply:
I went a bit through the paper of Elkan: http://web.cs.iastate.edu/~honavar/elkan.pdf
In section 2, there is an analytic formulation of the optimal threshold. However, it requires to use `predict_proba` (which could be fine since `method="auto"` by default?).
In this case (and I understand correctly), it means that you can compute the threshold just by looking at the cost matrix.

Reply:
Otherwise, your additive parametrization if fine with me, it makes sense.

Reply:
The Elkan paper also mentions the case of non-constant costs, that is costs that depends on the value of an input feature. This is practically very useful, probably much more useful than constant costs. Maybe we should re-scope this whole cost matrix thingy for another PR though. It will be much easier to implement once we have a clean API to deal with feature names.